donkeycar/templates/train.py

```diff
@@ -1,1043 +1,191 @@
-    index = sample['index']
-    return tub_path + str(index)
-
-
-def make_next_key(sample, index_offset):
-    tub_path = sample['tub_path']
-    index = sample['index'] + index_offset
-    return tub_path + str(index)
-
-
-def collate_records(records, gen_records, opts):
+class TubDataset(object):
     '''
-    open all the .json records from records list passed in,
-    read their contents,
-    add them to a list of gen_records, passed in.
-    use the opts dict to specify config choices
+    Loads the dataset, and creates a train/test split.
     '''
 
-    new_records = {}
-    
-    for record_path in records:
-
-        basepath = os.path.dirname(record_path)        
-        index = get_record_index(record_path)
-        sample = { 'tub_path' : basepath, "index" : index }
-             
-        key = make_key(sample)
-
-        if key in gen_records:
-            continue
-
-        try:
-            with open(record_path, 'r') as fp:
-                json_data = json.load(fp)
-        except:
-            continue
-
-        image_filename = json_data["cam/image_array"]
-        image_path = os.path.join(basepath, image_filename)
-
-        sample['record_path'] = record_path
-        sample["image_path"] = image_path
-        sample["json_data"] = json_data        
-
-        angle = float(json_data['user/angle'])
-        throttle = float(json_data["user/throttle"])
-
-        if opts['categorical']:
-            angle = dk.utils.linear_bin(angle)
-            throttle = dk.utils.linear_bin(throttle, N=20, offset=0, R=opts['cfg'].MODEL_CATEGORICAL_MAX_THROTTLE_RANGE)
-
-        sample['angle'] = angle
-        sample['throttle'] = throttle
-
-        try:
-            accl_x = float(json_data['imu/acl_x'])
-            accl_y = float(json_data['imu/acl_y'])
-            accl_z = float(json_data['imu/acl_z'])
-
-            gyro_x = float(json_data['imu/gyr_x'])
-            gyro_y = float(json_data['imu/gyr_y'])
-            gyro_z = float(json_data['imu/gyr_z'])
-
-            sample['imu_array'] = np.array([accl_x, accl_y, accl_z, gyro_x, gyro_y, gyro_z])
-        except:
-            pass
-
-        try:
-            behavior_arr = np.array(json_data['behavior/one_hot_state_array'])
-            sample["behavior_arr"] = behavior_arr
-        except:
-            pass
+    def __init__(self, tub_paths, test_size=0.2, shuffle=True):
+        self.tub_paths = tub_paths
+        self.test_size = test_size
+        self.shuffle = shuffle
+        self.tubs = [Tub(tub_path) for tub_path in self.tub_paths]
+        self.records = list()
+
+    def train_test_split(self):
+        print('Loading tubs from paths %s' % (self.tub_paths))
+        for tub in self.tubs:
+            for record in tub:
+                record['_image_base_path'] = tub.images_base_path
+                self.records.append(record)
+
+        return train_test_split(self.records, shuffle=self.shuffle, test_size=self.test_size)
+
+
+class TubSequence(Sequence):
```

Comment:
This could be moved into the Tub module. We could also generalise it by passing the arguments which data to pass for `X` and `y` into the constructor which would help to train different models that have more/other input data, like IMU, or output, like `KerasLatent`.

Reply:
Just a thought.